lightning/src/routing/router.rs

```diff
@@ -4968,6 +4994,8 @@ mod benches {
 				let dst = PublicKey::from_slice(nodes.keys().skip(seed % nodes.len()).next().unwrap().as_slice()).unwrap();
 				let payee = Payee::from_node_id(dst).with_features(InvoiceFeatures::known());
 				let amt = seed as u64 % 1_000_000;
+				let params = ProbabilisticScoringParameters::default();
+				let scorer = ProbabilisticScorer::new(params, &src, &graph);
```

Comment:
Can you add new benchmarks - one with the classic `Scorer` with no data (or, better yet, dummy scorer that always returns 0 that can be inlined), one with a `Scorer` with random data, and one with `ProbabilisticScorer`. In the interest of expediting this PR we can skip the second because the first just requires copying the function and maybe a five-line struct definition.

Reply:
Is it worth keeping around both `generate_routes` and `generate_mpp_routes` variations of the benchmarks? AFAICT, `get_route` will use MPP if node announcement  in the `NetworkGraph` indicates support, even if the invoice doesn't. Looking at some results, there isn't a material difference between the two for a given scorer.

```
test routing::router::benches::generate_mpp_routes_with_default_scorer       ... bench:  92,654,109 ns/iter (+/- 31,531,162)
test routing::router::benches::generate_mpp_routes_with_probabilistic_scorer ... bench: 107,804,711 ns/iter (+/- 53,006,509)
test routing::router::benches::generate_mpp_routes_with_zero_penalty_scorer  ... bench:  75,926,475 ns/iter (+/- 37,127,351)
test routing::router::benches::generate_routes_with_default_scorer           ... bench:  92,888,708 ns/iter (+/- 32,657,001)
test routing::router::benches::generate_routes_with_probabilistic_scorer     ... bench: 107,996,503 ns/iter (+/- 51,133,247)
test routing::router::benches::generate_routes_with_zero_penalty_scorer      ... bench:  75,003,204 ns/iter (+/- 36,967,684)
```

Reply:
Ah, seems I can hit the non-MPP case by using `Some(InvoiceFeatures::empty())` instead of `None`.

Added some commits up front that:
- Refactors out a benchmark utility to use with different scorers and features
- Exercises `first_hops`, which simplifies testing `ProbabilisticScorer`
- Seeds the scorer with success and failure data
- Benchmarks a zero-penalty scorer

The commits are fairly self-contained, but let me know if you prefer that I pull them out into a separate PR.